e2e/mandelbrot/mandelbrot.py

```diff
@@ -23,36 +23,63 @@ def mandelbrot(xmin, xmax, ymin, ymax, xn, yn, maxiter, horizon=2.0):
     N = np.zeros(C.shape, dtype=int)
     Z = np.zeros(C.shape, np.complex64)
     for n in range(maxiter):
-        I = np.less(abs(Z), horizon)
-        N[I] = n
-        Z[I] = Z[I]**2 + C[I]
-    N[N == maxiter-1] = 0
+        I = np.abs(Z) < horizon
+        N = np.where(I, n, N)  # N[I] = n
+        Z = np.where(I, Z**2 + C, Z)  # Z[I] = Z[I]**2 + C[I]
+    N = np.where(N == maxiter -1, 0, N)  # N[N == maxiter-1] = 0
     return Z, N
 
 
-if __name__ == '__main__':
-    from matplotlib import colors
-    import matplotlib.pyplot as plt
-  ##  from timeit import timeit
+def abs2(a):
+    r"""abs(a) replacement"""
+    return a[..., 0]**2 + a[..., 1]**2
 
-    # Benchmark
-    xmin, xmax, xn = -2.25, +0.75, int(3000/3)
-    ymin, ymax, yn = -1.25, +1.25, int(2500/3)
-    maxiter = 200
- ##   timeit("mandelbrot_1(xmin, xmax, ymin, ymax, xn, yn, maxiter)", globals())
- ##   timeit("mandelbrot_2(xmin, xmax, ymin, ymax, xn, yn, maxiter)", globals())
- ##   timeit("mandelbrot_3(xmin, xmax, ymin, ymax, xn, yn, maxiter)", globals())
 
+def sq2(a):
+    """a**2 replacement."""
+    z = np.empty_like(a)
+    z[..., 0] = a[..., 0]**2 - a[..., 1]**2
+    z[..., 1] = 2 * a[..., 0] * a[..., 1]
+    return z
+
+
+@torch.compile
+def step(n, CC, Z, N):
+    I = abs2(Z) < horizon**2
+    N = np.where(I, n, N)                       # N[I] = n
```

Comment:
This still graph breaks:

```
[2023-07-16 23:41:32,309] torch._dynamo.symbolic_convert: [DEBUG] TRACE starts_line mandelbrot.py:49
[2023-07-16 23:41:32,309] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_GLOBAL np []
[2023-07-16 23:41:32,309] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR where [PythonModuleVariable()]
[2023-07-16 23:41:32,309] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST I [NumpyVariable()]
[2023-07-16 23:41:32,309] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST n [NumpyVariable(), ConstantVariable(ndarray)]
[2023-07-16 23:41:32,309] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST N [NumpyVariable(), ConstantVariable(ndarray), ConstantVariable(int)]
[2023-07-16 23:41:32,309] torch._dynamo.symbolic_convert: [DEBUG] TRACE CALL_FUNCTION 3 [NumpyVariable(), ConstantVariable(ndarray), ConstantVariable(int), NumpyVariable()]
[2023-07-16 23:41:32,310] torch._dynamo.symbolic_convert: [DEBUG] break_graph_if_unsupported triggered compile
Traceback (most recent call last):
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 391, in wrapper
    return inner_fn(self, inst)
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 1100, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 559, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/variables/misc.py", line 809, in call_function
    unimplemented(f"numpy.{self.value}()")
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/exc.py", line 141, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: numpy.<function where at 0x7fb879455550>()
[2023-07-16 23:41:32,310] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: numpy.<function where at 0x7fb879455550>() from user code at   File "mandelbrot.py", line 49, in step
    N = np.where(I, n, N)                       # N[I] = n

[2023-07-16 23:41:32,310] torch._dynamo.output_graph: [DEBUG] restore_graphstate: removed 0 nodes
[2023-07-16 23:41:32,310] torch._dynamo.output_graph: [DEBUG] COMPILING GRAPH due to GraphCompileReason(reason='numpy.<function where at 0x7fb879455550>()', user_stack=[<FrameSummary file mandelbrot.py, line 49 in step>], graph_break=True)
[2023-07-16 23:41:32,311] torch._dynamo.symbolic_convert: [DEBUG] step triggered compile
Traceback (most recent call last):
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/codegen.py", line 153, in __call__
    output.extend(value.reconstruct(self))
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/variables/base.py", line 285, in reconstruct
    raise NotImplementedError()
NotImplementedError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 668, in step
    getattr(self, inst.opname)(inst)
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 437, in wrapper
    self.output.compile_subgraph(self, reason=reason)
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/output_graph.py", line 785, in compile_subgraph
    pass1.foreach(stack_values)
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/codegen.py", line 171, in foreach
    self(i)
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/codegen.py", line 155, in __call__
    unimplemented(f"reconstruct: {value}")
  File "/home/br/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/exc.py", line 141, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: reconstruct: ConstantVariable(ndarray)
```

However, trying to MWE this compiles fine:

```
import numpy as np

import torch
torch.set_default_device("cpu")
import torch._dynamo.config as cfg
cfg.numpy_ndarray_as_tensor = True


@torch.compile
def wh(n, N):
    z = np.where(N == n, n-1, N)
    return z


def loop(N, maxiter=5):
    for n in range(maxiter):
        N = wh(n, N)


N = np.arange(8)
loop(N)
```

Would appreciate a pointer @lezcano?

Reply:
In fact, the reason was just a missing `cfg.numpy_ndarray_as_tensor = True`. So, one more example of "compiler bugs exist surely but pebkac is more likely"